File: packages/sdk-ruby/lib/stagehand/webmcp.rb
# frozen_string_literal: true

require_relative "generated/models"
require_relative "rpc_client"

module Stagehand
  # A WebMCP tool advertised by the page, discovered via Page#tools. Port of
  # packages/sdk-python/src/stagehand/webmcp.py.
  class WebMCPTool
    attr_reader :name, :description, :input_schema, :annotations, :frame_id, :backend_node_id

    def initialize(rpc_client, page_id, descriptor)
      @rpc_client = rpc_client
      @page_id = page_id
      @name = descriptor.name
      @description = descriptor.description
      @input_schema = descriptor.input_schema
      @annotations = descriptor.annotations
      @frame_id = descriptor.frame_id
      @backend_node_id = descriptor.backend_node_id
    end

    # Starts the tool with the given input Hash and returns the running
    # WebMCPInvocation (await it with #result).
    def invoke(input: nil)
      descriptor = @rpc_client.send(
        "page.webmcp_invoke_tool",
        Models::PageWebMCPInvokeToolParams.new(
          page_id: @page_id,
          frame_id: @frame_id,
          tool_name: @name,
          input: input.nil? ? {} : input,
        ),
        "WebMCPInvocationDescriptor",
      )
      WebMCPInvocation.new(@rpc_client, @page_id, descriptor)
    end
  end

  # A running (or finished) WebMCP tool invocation.
  class WebMCPInvocation
    attr_reader :invocation_id, :tool_name, :frame_id, :input

    def initialize(rpc_client, page_id, descriptor)
      @rpc_client = rpc_client
      @page_id = page_id
      @invocation_id = descriptor.invocation_id
      @tool_name = descriptor.tool_name
      @frame_id = descriptor.frame_id
      @input = descriptor.input
      @terminal_result = nil
    end

    # Blocks until the invocation reaches a terminal state and returns the
    # Models::WebMCPToolResponse (status Completed/Canceled/Error). The
    # terminal result is memoized; later calls do not re-issue the RPC.
    def result(timeout: nil)
      return @terminal_result unless @terminal_result.nil?

      values = { page_id: @page_id, invocation_id: @invocation_id }
      values[:options] = Models::WebMCPResultOptions.new(timeout: timeout) unless timeout.nil?
      @terminal_result = @rpc_client.send(
        "page.webmcp_invocation_result",
        Models::PageWebMCPInvocationResultParams.new(**values),
        "WebMCPToolResponse",
      )
    end

    def cancel
      @rpc_client.send(
        "page.webmcp_cancel_invocation",
        Models::PageWebMCPCancelInvocationParams.new(page_id: @page_id, invocation_id: @invocation_id),
        "PageVoidResult",
      )
      nil
    end
  end
end

File: packages/sdk-go/webmcp.go
package stagehand

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
)

// WebMCPInput is JSON input passed to a page-provided WebMCP tool.
type WebMCPInput map[string]any

// WebMCPTool is a page-bound wrapper around a discovered WebMCP tool.
type WebMCPTool struct {
	rpc        protocolClient
	pageID     string
	descriptor WebMCPToolDescriptor
}

// Descriptor returns the generated wire descriptor for the tool.
func (t *WebMCPTool) Descriptor() WebMCPToolDescriptor {
	return t.descriptor
}

// Invoke invokes the tool using the page, frame, and name it already owns.
// A nil input is sent as an empty JSON object.
func (t *WebMCPTool) Invoke(
	ctx context.Context,
	input WebMCPInput,
) (*WebMCPInvocation, error) {
	wireInput, err := encodeWebMCPInput(input)
	if err != nil {
		return nil, err
	}

	params := PageWebMCPInvokeToolParams{
		PageID:   t.pageID,
		FrameID:  t.descriptor.FrameID,
		ToolName: t.descriptor.Name,
		Input:    wireInput,
	}
	var descriptor WebMCPInvocationDescriptor
	if err := t.rpc.call(ctx, "page.webmcp_invoke_tool", params, &descriptor); err != nil {
		return nil, err
	}
	return &WebMCPInvocation{
		rpc:        t.rpc,
		pageID:     t.pageID,
		descriptor: descriptor,
	}, nil
}

// WebMCPInvocation is a page-bound handle for an invocation accepted by Chrome.
type WebMCPInvocation struct {
	rpc        protocolClient
	pageID     string
	descriptor WebMCPInvocationDescriptor

	mu             sync.RWMutex
	terminalResult *WebMCPToolResponse
}

// Descriptor returns the generated wire descriptor for the invocation.
func (i *WebMCPInvocation) Descriptor() WebMCPInvocationDescriptor {
	return i.descriptor
}

// Result waits for Chrome's authoritative terminal invocation response.
// Successful waits are cached; context and RPC failures can be retried.
func (i *WebMCPInvocation) Result(
	ctx context.Context,
	options *WebMCPResultOptions,
) (WebMCPToolResponse, error) {
	i.mu.RLock()
	cached := i.terminalResult
	i.mu.RUnlock()
	if cached != nil {
		return *cached, nil
	}

	params := PageWebMCPInvocationResultParams{
		PageID:       i.pageID,
		InvocationID: i.descriptor.InvocationID,
		Options:      options,
	}
	var result WebMCPToolResponse
	if err := i.rpc.call(ctx, "page.webmcp_invocation_result", params, &result); err != nil {
		return WebMCPToolResponse{}, err
	}

	i.mu.Lock()
	if i.terminalResult == nil {
		i.terminalResult = &result
	}
	cached = i.terminalResult
	i.mu.Unlock()
	return *cached, nil
}

// Cancel requests cancellation without changing the invocation's terminal result locally.
func (i *WebMCPInvocation) Cancel(ctx context.Context) error {
	params := PageWebMCPCancelInvocationParams{
		PageID:       i.pageID,
		InvocationID: i.descriptor.InvocationID,
	}
	var result PageVoidResult
	return i.rpc.call(ctx, "page.webmcp_cancel_invocation", params, &result)
}

// WebMCPOutputAs decodes a terminal response's JSON output into a caller-selected Go type.
func WebMCPOutputAs[T any](response WebMCPToolResponse) (T, error) {
	var output T
	if len(response.Output) == 0 {
		return output, fmt.Errorf("WebMCP response %q has no output", response.InvocationID)
	}
	if err := json.Unmarshal(response.Output, &output); err != nil {
		return output, fmt.Errorf("decode WebMCP response %q output: %w", response.InvocationID, err)
	}
	return output, nil
}

func encodeWebMCPInput(input WebMCPInput) (PageWebMCPInvokeToolParamsInput, error) {
	encoded := make(PageWebMCPInvokeToolParamsInput, len(input))
	for key, value := range input {
		raw, err := json.Marshal(value)
		if err != nil {
			return nil, fmt.Errorf("encode WebMCP input %q: %w", key, err)
		}
		encoded[key] = raw
	}
	return encoded, nil
}
